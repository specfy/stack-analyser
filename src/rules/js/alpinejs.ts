import { register } from '../../register.js';

register({
  tech: 'alpinejs',
  name: 'Alpine.js',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'alpinejs' }],
});
