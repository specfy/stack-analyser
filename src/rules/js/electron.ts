import { register } from '../../register.js';

register({
  tech: 'electron',
  name: 'Electron',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'electron' }],
});
