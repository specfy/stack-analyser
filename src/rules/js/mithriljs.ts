import { register } from '../../register.js';

register({
  tech: 'mithriljs',
  name: 'Mithril',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'mithril' }],
});
