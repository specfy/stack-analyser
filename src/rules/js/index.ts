import './alpinejs.js';
import './angular.js';
import './bootstrap.js';
import './emberjs.js';
import './esbuild.js';
import './eslint.js';
import './express.js';
import './fastify.js';
import './koa.js';
import './meteorjs.js';
import './mithriljs.js';
import './nextjs.js';
import './nuxtjs.js';
import './preactjs.js';
import './prettier.js';
import './prisma.js';
import './puppeteer.js';
import './react.js';
import './reactEmail.js';
import './rollup.js';
import './sequelize.js';
import './socketio.js';
import './solidjs.js';
import './storybook.js';
import './strapi.js';
import './stylelint.js';
import './sveltejs.js';
import './tailwind.js';
import './typescript.js';
import './vite.js';
import './vue.js';
import './webpack.js';
