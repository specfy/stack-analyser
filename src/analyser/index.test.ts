import path from 'node:path';

import { describe, it, expect, vi, beforeEach } from 'vitest';

import { flatten } from '../payload/helpers.js';
import { FakeProvider } from '../provider/fake.js';
import { FSProvider } from '../provider/fs.js';
import type { AnalyserJson } from '../types/index.js';

import { analyser } from './index.js';

import '../autoload.js';

const dockerCompose = `version: '3'
services:
  db:
    container_name: db
    image: postgres:15.1-alpine
    ports:
      - '5432:5432'
    environment:
      - POSTGRES_PASSWORD=postgres
`;

let id = 0;
vi.mock('../common/nid.ts', () => {
  return { nid: () => `${id++}` };
});

describe('analyser', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('should not find anything', async () => {
    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': [],
        },
        files: {},
      }),
    });

    expect(res.toJson('')).toStrictEqual({
      id: expect.any(String),
      name: 'main',
      edges: [],
      inComponent: null,
      languages: {},
      path: ['/'],
      tech: null,
      techs: [],
      childs: [],
      dependencies: [],
      reason: [],
    });
  });

  it('should register only component of the same tech', async () => {
    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['package.json', 'docker-compose.yml'],
        },
        files: {
          '/docker-compose.yml': dockerCompose,
          '/package.json': '{ "name": "test", "dependencies": {"pg": "1.0.0"}}',
        },
      }),
    });

    const flat = flatten(res, { merge: true });
    const json: AnalyserJson = JSON.parse(JSON.stringify(flat.toJson('')));
    expect(json).toMatchSnapshot();
    expect(flat.childs[0].id).toBe(flat.childs[1].edges[0].target.id);
  });

  it('should run correctly', async () => {
    const root = path.join(__dirname, '../../tests/__fixtures__');
    const res = await analyser({
      provider: new FSProvider({
        path: root,
      }),
    });

    expect(res.toJson(root)).toMatchSnapshot();

    const flatted = flatten(res, { merge: true });

    // Check that inComponent was updated
    const vercel = flatted.childs.find((child) => child.name === 'Vercel')!;
    const app = flatted.childs.find((child) => child.name === '@fake/app');
    expect(app!.inComponent!.id).toBe(vercel.id);

    // Check that edge.target was updated
    const datadog = flatted.childs.find((child) => child.name === 'Datadog')!;
    const api = flatted.childs.find((child) => child.name === '@fake/api');
    expect(api!.edges[0].target.id).toBe(datadog.id);

    expect(JSON.parse(JSON.stringify(flatted.toJson(root)))).toMatchSnapshot();
  });
});
