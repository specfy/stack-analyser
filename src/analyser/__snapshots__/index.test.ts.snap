// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyser > should register only component of the same tech 1`] = `
{
  "childs": [
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "postgres",
          "15.1-alpine",
        ],
      ],
      "edges": [],
      "id": "3",
      "inComponent": null,
      "languages": {},
      "name": "db",
      "path": [
        "/docker-compose.yml",
        "/package.json",
      ],
      "reason": [],
      "tech": "postgresql",
      "techs": [
        "postgresql",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "pg",
          "1.0.0",
        ],
      ],
      "edges": [
        {
          "read": true,
          "target": "3",
          "write": true,
        },
      ],
      "id": "4",
      "inComponent": null,
      "languages": {
        "JSON": 1,
        "YAML": 1,
      },
      "name": "test",
      "path": [
        "/package.json",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "docker",
        "nodejs",
        "postgresql",
      ],
    },
  ],
  "dependencies": [
    [
      "docker",
      "postgres",
      "15.1-alpine",
    ],
    [
      "npm",
      "pg",
      "1.0.0",
    ],
  ],
  "edges": [],
  "id": "6",
  "inComponent": null,
  "languages": {
    "JSON": 2,
    "YAML": 2,
  },
  "name": "flatten",
  "path": [
    "/",
    "/docker-compose.yml",
    "/package.json",
  ],
  "reason": [],
  "tech": null,
  "techs": [
    "docker",
    "nodejs",
    "postgresql",
  ],
}
`;

exports[`analyser > should run correctly 1`] = `
{
  "childs": [
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "postgres",
          "14.5-alpine",
        ],
      ],
      "edges": [],
      "id": "9",
      "inComponent": null,
      "languages": {},
      "name": "db",
      "path": [
        "/docker-compose.yml",
      ],
      "reason": [
        "postgresql matched: /postgres/",
      ],
      "tech": "postgresql",
      "techs": [],
    },
    {
      "childs": [
        {
          "childs": [
            {
              "childs": [],
              "dependencies": [],
              "edges": [],
              "id": "21",
              "inComponent": null,
              "languages": {},
              "name": "Datadog",
              "path": [
                "/pkgs/api/package.json",
              ],
              "reason": [
                "datadog matched: /^dd-trace$/",
              ],
              "tech": "datadog",
              "techs": [],
            },
          ],
          "dependencies": [
            [
              "npm",
              "dd-trace",
              "1.0.1",
            ],
            [
              "npm",
              "fastify",
              "4.17.0",
            ],
            [
              "npm",
              "prisma",
              "4.13.0",
            ],
            [
              "npm",
              "typescript",
              "5.0.4",
            ],
          ],
          "edges": [
            {
              "read": true,
              "target": "21",
              "write": true,
            },
          ],
          "id": "20",
          "inComponent": null,
          "languages": {
            "JSON": 1,
          },
          "name": "@fake/api",
          "path": [
            "/pkgs/api/package.json",
          ],
          "reason": [
            "datadog matched: /^dd-trace$/",
            "fastify matched: /^fastify$/",
            "prisma matched: /^prisma$/",
            "typescript matched: /^typescript$/",
            "matched file: package.json",
          ],
          "tech": null,
          "techs": [
            "datadog",
            "fastify",
            "nodejs",
            "prisma",
            "typescript",
          ],
        },
        {
          "childs": [
            {
              "childs": [],
              "dependencies": [],
              "edges": [],
              "id": "23",
              "inComponent": null,
              "languages": {},
              "name": "Vercel",
              "path": [
                "/pkgs/app/package.json",
              ],
              "reason": [
                "matched file: .vercel",
              ],
              "tech": "vercel",
              "techs": [],
            },
          ],
          "dependencies": [
            [
              "npm",
              "@fake/api",
              "1.0.0",
            ],
            [
              "npm",
              "react",
              "4.17.0",
            ],
            [
              "npm",
              "typescript",
              "5.0.4",
            ],
            [
              "npm",
              "vite",
              "4.3.3",
            ],
          ],
          "edges": [],
          "id": "22",
          "inComponent": "23",
          "languages": {
            "HTML": 1,
            "JSON": 1,
            "SCSS": 1,
          },
          "name": "@fake/app",
          "path": [
            "/pkgs/app/package.json",
          ],
          "reason": [
            "react matched: /^react$/",
            "typescript matched: /^typescript$/",
            "vite matched: /^vite$/",
            "matched file: .vercel",
            "matched file: package.json",
          ],
          "tech": null,
          "techs": [
            "nodejs",
            "react",
            "typescript",
            "vercel",
            "vite",
          ],
        },
        {
          "childs": [],
          "dependencies": [
            [
              "terraform",
              "registry.terraform.io/hashicorp/google",
              "4.61.0",
            ],
          ],
          "edges": [],
          "id": "27",
          "inComponent": null,
          "languages": {},
          "name": "Google Cloud Platform",
          "path": [
            "/terraform/.terraform.lock.hcl",
          ],
          "reason": [
            "gcp matched: /^registry.terraform.io\\\\/hashicorp\\\\/google$/",
          ],
          "tech": "gcp",
          "techs": [],
        },
        {
          "childs": [],
          "dependencies": [],
          "edges": [],
          "id": "14",
          "inComponent": null,
          "languages": {},
          "name": "Postgres",
          "path": [
            "/deno/deno.lock",
          ],
          "reason": [
            "postgresql matched: /\\\\/x\\\\/postgres@/",
          ],
          "tech": "postgresql",
          "techs": [],
        },
        {
          "childs": [],
          "dependencies": [
            [
              "terraform",
              "registry.terraform.io/vercel/vercel",
              "4.61.0",
            ],
          ],
          "edges": [],
          "id": "28",
          "inComponent": null,
          "languages": {},
          "name": "Vercel",
          "path": [
            "/terraform/.terraform.lock.hcl",
          ],
          "reason": [
            "vercel matched: /^registry.terraform.io\\\\/vercel\\\\/vercel$/",
          ],
          "tech": "vercel",
          "techs": [],
        },
        {
          "childs": [],
          "dependencies": [
            [
              "rust",
              "dotenv",
              "0.15.0",
            ],
          ],
          "edges": [],
          "id": "25",
          "inComponent": null,
          "languages": {
            "TOML": 1,
          },
          "name": "rust-server",
          "path": [
            "/rust/Cargo.toml",
          ],
          "reason": [
            "matched file: Cargo.toml",
          ],
          "tech": null,
          "techs": [
            "rust",
          ],
        },
        {
          "childs": [],
          "dependencies": [],
          "edges": [],
          "id": "16",
          "inComponent": null,
          "languages": {
            "Go": 1,
          },
          "name": "server",
          "path": [
            "/golang/cmd/server/main.go",
          ],
          "reason": [
            "matched file: main.go",
            "matched file: main.go",
          ],
          "tech": "golang",
          "techs": [
            "golang",
            "rust",
          ],
        },
        {
          "childs": [
            {
              "childs": [],
              "dependencies": [],
              "edges": [],
              "id": "19",
              "inComponent": null,
              "languages": {},
              "name": "Algolia",
              "path": [
                "/php/composer.json",
              ],
              "reason": [
                "algolia matched: /^algolia\\\\/algoliasearch-client-php$/",
              ],
              "tech": "algolia",
              "techs": [],
            },
            {
              "childs": [],
              "dependencies": [],
              "edges": [],
              "id": "18",
              "inComponent": null,
              "languages": {},
              "name": "Redis",
              "path": [
                "/php/composer.json",
              ],
              "reason": [
                "redis matched: /^predis\\\\/predis$/",
              ],
              "tech": "redis",
              "techs": [],
            },
          ],
          "dependencies": [
            [
              "php",
              "mockery/mockery",
              "~1.0",
            ],
            [
              "php",
              "phpunit/phpunit",
              "^9.0",
            ],
            [
              "php",
              "php-mock/php-mock-mockery",
              "^1.3",
            ],
            [
              "php",
              "dms/phpunit-arraysubset-asserts",
              "^0.2.1",
            ],
            [
              "php",
              "fakerphp/faker",
              "^1.13",
            ],
            [
              "php",
              "slevomat/coding-standard",
              "^7.0",
            ],
            [
              "php",
              "nunomaduro/larastan",
              "^2.1",
            ],
            [
              "php",
              "laravel/tinker",
              "^2.7",
            ],
            [
              "php",
              "php",
              ">=8.0",
            ],
            [
              "php",
              "laravel/framework",
              "^9.0",
            ],
            [
              "php",
              "james-heinrich/getid3",
              "^1.9",
            ],
            [
              "php",
              "guzzlehttp/guzzle",
              "^7.0.1",
            ],
            [
              "php",
              "aws/aws-sdk-php-laravel",
              "^3.1",
            ],
            [
              "php",
              "pusher/pusher-php-server",
              "^4.0",
            ],
            [
              "php",
              "predis/predis",
              "~1.0",
            ],
            [
              "php",
              "jackiedo/dotenv-editor",
              "^2.0",
            ],
            [
              "php",
              "ext-exif",
              "*",
            ],
            [
              "php",
              "ext-gd",
              "*",
            ],
            [
              "php",
              "ext-fileinfo",
              "*",
            ],
            [
              "php",
              "ext-json",
              "*",
            ],
            [
              "php",
              "ext-SimpleXML",
              "*",
            ],
            [
              "php",
              "daverandom/resume",
              "^0.0.3",
            ],
            [
              "php",
              "laravel/helpers",
              "^1.0",
            ],
            [
              "php",
              "intervention/image",
              "^2.5",
            ],
            [
              "php",
              "doctrine/dbal",
              "^3.0",
            ],
            [
              "php",
              "lstrojny/functional-php",
              "^1.14",
            ],
            [
              "php",
              "teamtnt/laravel-scout-tntsearch-driver",
              "^11.1",
            ],
            [
              "php",
              "algolia/algoliasearch-client-php",
              "^3.3",
            ],
            [
              "php",
              "laravel/ui",
              "^3.2",
            ],
            [
              "php",
              "webmozart/assert",
              "^1.10",
            ],
            [
              "php",
              "laravel/sanctum",
              "^2.15",
            ],
            [
              "php",
              "laravel/scout",
              "^9.4",
            ],
            [
              "php",
              "nunomaduro/collision",
              "^6.2",
            ],
            [
              "php",
              "jwilsson/spotify-web-api-php",
              "^5.2",
            ],
            [
              "php",
              "meilisearch/meilisearch-php",
              "^0.24.0",
            ],
            [
              "php",
              "http-interop/http-factory-guzzle",
              "^1.2",
            ],
          ],
          "edges": [
            {
              "read": true,
              "target": "18",
              "write": true,
            },
            {
              "read": true,
              "target": "19",
              "write": true,
            },
          ],
          "id": "17",
          "inComponent": null,
          "languages": {
            "JSON": 1,
          },
          "name": "symfony/translation",
          "path": [
            "/php/composer.json",
          ],
          "reason": [
            "phpunit matched: /^phpunit\\\\/phpunit$/",
            "laravel matched: /^laravel\\\\/framework$/",
            "redis matched: /^predis\\\\/predis$/",
            "algolia matched: /^algolia\\\\/algoliasearch-client-php$/",
            "matched file: composer.json",
          ],
          "tech": null,
          "techs": [
            "algolia",
            "laravel",
            "php",
            "phpunit",
            "redis",
          ],
        },
      ],
      "dependencies": [
        [
          "npm",
          "@typescript-eslint/eslint-plugin",
          "5.57.1",
        ],
        [
          "npm",
          "@typescript-eslint/parser",
          "5.57.1",
        ],
        [
          "npm",
          "eslint",
          "8.39.0",
        ],
        [
          "npm",
          "prettier",
          "2.8.7",
        ],
        [
          "npm",
          "typescript",
          "4.9.5",
        ],
        [
          "deno",
          "https://deno.land/x/postgres@v0.17.0/mod.ts",
          "23a1a36928f1b6d3b0170aaa67de09af12aa998525f608ff7331b9fb364cbde6",
        ],
        [
          "golang",
          "github.com/google/uuid",
          "v1.3.0",
        ],
        [
          "ruby",
          "rails",
          "6.1.7.4",
        ],
        [
          "ruby",
          "puma\\", \\"6.3.0",
          "latest",
        ],
        [
          "terraform",
          "registry.terraform.io/hashicorp/google",
          "4.61.0",
        ],
        [
          "terraform",
          "registry.terraform.io/vercel/vercel",
          "4.61.0",
        ],
      ],
      "edges": [],
      "id": "12",
      "inComponent": null,
      "languages": {
        "AMPL": 1,
        "HCL": 1,
        "JSON": 1,
        "YAML": 1,
      },
      "name": "fake",
      "path": [
        "/package.json",
      ],
      "reason": [
        "eslint matched: /^eslint$/",
        "prettier matched: /^prettier$/",
        "typescript matched: /^typescript$/",
        "matched file: docker-compose.yml",
        "matched file: package.json",
        "matched file: deno.lock",
        "matched file: go.mod",
        "matched file: Gemfile",
        "matched file: .terraform.lock.hcl",
      ],
      "tech": null,
      "techs": [
        "deno",
        "docker",
        "eslint",
        "golang",
        "nodejs",
        "prettier",
        "ruby",
        "terraform",
        "typescript",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "redis",
          "7.0.4-alpine",
        ],
      ],
      "edges": [],
      "id": "10",
      "inComponent": null,
      "languages": {},
      "name": "redis",
      "path": [
        "/docker-compose.yml",
      ],
      "reason": [
        "redis matched: /redis/",
      ],
      "tech": "redis",
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "unknown",
          "7.17.5",
        ],
      ],
      "edges": [],
      "id": "11",
      "inComponent": null,
      "languages": {},
      "name": "unknown",
      "path": [
        "/docker-compose.yml",
      ],
      "reason": [
        "matched: unknown",
      ],
      "tech": null,
      "techs": [],
    },
  ],
  "dependencies": [],
  "edges": [],
  "id": "7",
  "inComponent": null,
  "languages": {},
  "name": "main",
  "path": [
    "/",
  ],
  "reason": [],
  "tech": null,
  "techs": [],
}
`;

exports[`analyser > should run correctly 2`] = `
{
  "childs": [
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "dd-trace",
          "1.0.1",
        ],
        [
          "npm",
          "fastify",
          "4.17.0",
        ],
        [
          "npm",
          "prisma",
          "4.13.0",
        ],
        [
          "npm",
          "typescript",
          "5.0.4",
        ],
      ],
      "edges": [
        {
          "read": true,
          "target": "21",
          "write": true,
        },
      ],
      "id": "20",
      "inComponent": null,
      "languages": {
        "JSON": 1,
      },
      "name": "@fake/api",
      "path": [
        "/pkgs/api/package.json",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "datadog",
        "fastify",
        "nodejs",
        "prisma",
        "typescript",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "@fake/api",
          "1.0.0",
        ],
        [
          "npm",
          "react",
          "4.17.0",
        ],
        [
          "npm",
          "typescript",
          "5.0.4",
        ],
        [
          "npm",
          "vite",
          "4.3.3",
        ],
      ],
      "edges": [
        {
          "read": true,
          "target": "20",
          "write": true,
        },
      ],
      "id": "22",
      "inComponent": "23",
      "languages": {
        "HTML": 1,
        "JSON": 1,
        "SCSS": 1,
      },
      "name": "@fake/app",
      "path": [
        "/pkgs/app/package.json",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "nodejs",
        "react",
        "typescript",
        "vercel",
        "vite",
      ],
    },
    {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "id": "19",
      "inComponent": null,
      "languages": {},
      "name": "Algolia",
      "path": [
        "/php/composer.json",
      ],
      "reason": [],
      "tech": "algolia",
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "id": "21",
      "inComponent": null,
      "languages": {},
      "name": "Datadog",
      "path": [
        "/pkgs/api/package.json",
      ],
      "reason": [],
      "tech": "datadog",
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "terraform",
          "registry.terraform.io/hashicorp/google",
          "4.61.0",
        ],
      ],
      "edges": [],
      "id": "27",
      "inComponent": null,
      "languages": {},
      "name": "Google Cloud Platform",
      "path": [
        "/terraform/.terraform.lock.hcl",
      ],
      "reason": [],
      "tech": "gcp",
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "terraform",
          "registry.terraform.io/vercel/vercel",
          "4.61.0",
        ],
      ],
      "edges": [],
      "id": "23",
      "inComponent": null,
      "languages": {},
      "name": "Vercel",
      "path": [
        "/pkgs/app/package.json",
        "/terraform/.terraform.lock.hcl",
      ],
      "reason": [],
      "tech": "vercel",
      "techs": [
        "vercel",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "postgres",
          "14.5-alpine",
        ],
      ],
      "edges": [],
      "id": "9",
      "inComponent": null,
      "languages": {},
      "name": "db",
      "path": [
        "/docker-compose.yml",
        "/deno/deno.lock",
      ],
      "reason": [],
      "tech": "postgresql",
      "techs": [
        "postgresql",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "npm",
          "@typescript-eslint/eslint-plugin",
          "5.57.1",
        ],
        [
          "npm",
          "@typescript-eslint/parser",
          "5.57.1",
        ],
        [
          "npm",
          "eslint",
          "8.39.0",
        ],
        [
          "npm",
          "prettier",
          "2.8.7",
        ],
        [
          "npm",
          "typescript",
          "4.9.5",
        ],
        [
          "deno",
          "https://deno.land/x/postgres@v0.17.0/mod.ts",
          "23a1a36928f1b6d3b0170aaa67de09af12aa998525f608ff7331b9fb364cbde6",
        ],
        [
          "golang",
          "github.com/google/uuid",
          "v1.3.0",
        ],
        [
          "ruby",
          "rails",
          "6.1.7.4",
        ],
        [
          "ruby",
          "puma\\", \\"6.3.0",
          "latest",
        ],
        [
          "terraform",
          "registry.terraform.io/hashicorp/google",
          "4.61.0",
        ],
        [
          "terraform",
          "registry.terraform.io/vercel/vercel",
          "4.61.0",
        ],
      ],
      "edges": [],
      "id": "12",
      "inComponent": null,
      "languages": {
        "AMPL": 1,
        "Go": 1,
        "HCL": 1,
        "HTML": 1,
        "JSON": 4,
        "SCSS": 1,
        "TOML": 1,
        "YAML": 1,
      },
      "name": "fake",
      "path": [
        "/package.json",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "deno",
        "docker",
        "eslint",
        "golang",
        "nodejs",
        "prettier",
        "ruby",
        "terraform",
        "typescript",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "redis",
          "7.0.4-alpine",
        ],
      ],
      "edges": [],
      "id": "10",
      "inComponent": null,
      "languages": {},
      "name": "redis",
      "path": [
        "/docker-compose.yml",
        "/php/composer.json",
      ],
      "reason": [],
      "tech": "redis",
      "techs": [
        "redis",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "rust",
          "dotenv",
          "0.15.0",
        ],
      ],
      "edges": [],
      "id": "25",
      "inComponent": null,
      "languages": {
        "TOML": 1,
      },
      "name": "rust-server",
      "path": [
        "/rust/Cargo.toml",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "rust",
      ],
    },
    {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "id": "16",
      "inComponent": null,
      "languages": {
        "Go": 1,
      },
      "name": "server",
      "path": [
        "/golang/cmd/server/main.go",
      ],
      "reason": [],
      "tech": "golang",
      "techs": [
        "golang",
        "rust",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "php",
          "mockery/mockery",
          "~1.0",
        ],
        [
          "php",
          "phpunit/phpunit",
          "^9.0",
        ],
        [
          "php",
          "php-mock/php-mock-mockery",
          "^1.3",
        ],
        [
          "php",
          "dms/phpunit-arraysubset-asserts",
          "^0.2.1",
        ],
        [
          "php",
          "fakerphp/faker",
          "^1.13",
        ],
        [
          "php",
          "slevomat/coding-standard",
          "^7.0",
        ],
        [
          "php",
          "nunomaduro/larastan",
          "^2.1",
        ],
        [
          "php",
          "laravel/tinker",
          "^2.7",
        ],
        [
          "php",
          "php",
          ">=8.0",
        ],
        [
          "php",
          "laravel/framework",
          "^9.0",
        ],
        [
          "php",
          "james-heinrich/getid3",
          "^1.9",
        ],
        [
          "php",
          "guzzlehttp/guzzle",
          "^7.0.1",
        ],
        [
          "php",
          "aws/aws-sdk-php-laravel",
          "^3.1",
        ],
        [
          "php",
          "pusher/pusher-php-server",
          "^4.0",
        ],
        [
          "php",
          "predis/predis",
          "~1.0",
        ],
        [
          "php",
          "jackiedo/dotenv-editor",
          "^2.0",
        ],
        [
          "php",
          "ext-exif",
          "*",
        ],
        [
          "php",
          "ext-gd",
          "*",
        ],
        [
          "php",
          "ext-fileinfo",
          "*",
        ],
        [
          "php",
          "ext-json",
          "*",
        ],
        [
          "php",
          "ext-SimpleXML",
          "*",
        ],
        [
          "php",
          "daverandom/resume",
          "^0.0.3",
        ],
        [
          "php",
          "laravel/helpers",
          "^1.0",
        ],
        [
          "php",
          "intervention/image",
          "^2.5",
        ],
        [
          "php",
          "doctrine/dbal",
          "^3.0",
        ],
        [
          "php",
          "lstrojny/functional-php",
          "^1.14",
        ],
        [
          "php",
          "teamtnt/laravel-scout-tntsearch-driver",
          "^11.1",
        ],
        [
          "php",
          "algolia/algoliasearch-client-php",
          "^3.3",
        ],
        [
          "php",
          "laravel/ui",
          "^3.2",
        ],
        [
          "php",
          "webmozart/assert",
          "^1.10",
        ],
        [
          "php",
          "laravel/sanctum",
          "^2.15",
        ],
        [
          "php",
          "laravel/scout",
          "^9.4",
        ],
        [
          "php",
          "nunomaduro/collision",
          "^6.2",
        ],
        [
          "php",
          "jwilsson/spotify-web-api-php",
          "^5.2",
        ],
        [
          "php",
          "meilisearch/meilisearch-php",
          "^0.24.0",
        ],
        [
          "php",
          "http-interop/http-factory-guzzle",
          "^1.2",
        ],
      ],
      "edges": [
        {
          "read": true,
          "target": "10",
          "write": true,
        },
        {
          "read": true,
          "target": "19",
          "write": true,
        },
      ],
      "id": "17",
      "inComponent": null,
      "languages": {
        "JSON": 1,
      },
      "name": "symfony/translation",
      "path": [
        "/php/composer.json",
      ],
      "reason": [],
      "tech": null,
      "techs": [
        "algolia",
        "laravel",
        "php",
        "phpunit",
        "redis",
      ],
    },
    {
      "childs": [],
      "dependencies": [
        [
          "docker",
          "unknown",
          "7.17.5",
        ],
      ],
      "edges": [],
      "id": "11",
      "inComponent": null,
      "languages": {},
      "name": "unknown",
      "path": [
        "/docker-compose.yml",
      ],
      "reason": [],
      "tech": null,
      "techs": [],
    },
  ],
  "dependencies": [
    [
      "docker",
      "postgres",
      "14.5-alpine",
    ],
    [
      "docker",
      "redis",
      "7.0.4-alpine",
    ],
    [
      "docker",
      "unknown",
      "7.17.5",
    ],
    [
      "php",
      "mockery/mockery",
      "~1.0",
    ],
    [
      "php",
      "phpunit/phpunit",
      "^9.0",
    ],
    [
      "php",
      "php-mock/php-mock-mockery",
      "^1.3",
    ],
    [
      "php",
      "dms/phpunit-arraysubset-asserts",
      "^0.2.1",
    ],
    [
      "php",
      "fakerphp/faker",
      "^1.13",
    ],
    [
      "php",
      "slevomat/coding-standard",
      "^7.0",
    ],
    [
      "php",
      "nunomaduro/larastan",
      "^2.1",
    ],
    [
      "php",
      "laravel/tinker",
      "^2.7",
    ],
    [
      "php",
      "php",
      ">=8.0",
    ],
    [
      "php",
      "laravel/framework",
      "^9.0",
    ],
    [
      "php",
      "james-heinrich/getid3",
      "^1.9",
    ],
    [
      "php",
      "guzzlehttp/guzzle",
      "^7.0.1",
    ],
    [
      "php",
      "aws/aws-sdk-php-laravel",
      "^3.1",
    ],
    [
      "php",
      "pusher/pusher-php-server",
      "^4.0",
    ],
    [
      "php",
      "predis/predis",
      "~1.0",
    ],
    [
      "php",
      "jackiedo/dotenv-editor",
      "^2.0",
    ],
    [
      "php",
      "ext-exif",
      "*",
    ],
    [
      "php",
      "ext-gd",
      "*",
    ],
    [
      "php",
      "ext-fileinfo",
      "*",
    ],
    [
      "php",
      "ext-json",
      "*",
    ],
    [
      "php",
      "ext-SimpleXML",
      "*",
    ],
    [
      "php",
      "daverandom/resume",
      "^0.0.3",
    ],
    [
      "php",
      "laravel/helpers",
      "^1.0",
    ],
    [
      "php",
      "intervention/image",
      "^2.5",
    ],
    [
      "php",
      "doctrine/dbal",
      "^3.0",
    ],
    [
      "php",
      "lstrojny/functional-php",
      "^1.14",
    ],
    [
      "php",
      "teamtnt/laravel-scout-tntsearch-driver",
      "^11.1",
    ],
    [
      "php",
      "algolia/algoliasearch-client-php",
      "^3.3",
    ],
    [
      "php",
      "laravel/ui",
      "^3.2",
    ],
    [
      "php",
      "webmozart/assert",
      "^1.10",
    ],
    [
      "php",
      "laravel/sanctum",
      "^2.15",
    ],
    [
      "php",
      "laravel/scout",
      "^9.4",
    ],
    [
      "php",
      "nunomaduro/collision",
      "^6.2",
    ],
    [
      "php",
      "jwilsson/spotify-web-api-php",
      "^5.2",
    ],
    [
      "php",
      "meilisearch/meilisearch-php",
      "^0.24.0",
    ],
    [
      "php",
      "http-interop/http-factory-guzzle",
      "^1.2",
    ],
    [
      "npm",
      "dd-trace",
      "1.0.1",
    ],
    [
      "npm",
      "fastify",
      "4.17.0",
    ],
    [
      "npm",
      "prisma",
      "4.13.0",
    ],
    [
      "npm",
      "typescript",
      "5.0.4",
    ],
    [
      "terraform",
      "registry.terraform.io/vercel/vercel",
      "4.61.0",
    ],
    [
      "npm",
      "@fake/api",
      "1.0.0",
    ],
    [
      "npm",
      "react",
      "4.17.0",
    ],
    [
      "npm",
      "vite",
      "4.3.3",
    ],
    [
      "rust",
      "dotenv",
      "0.15.0",
    ],
    [
      "terraform",
      "registry.terraform.io/hashicorp/google",
      "4.61.0",
    ],
    [
      "npm",
      "@typescript-eslint/eslint-plugin",
      "5.57.1",
    ],
    [
      "npm",
      "@typescript-eslint/parser",
      "5.57.1",
    ],
    [
      "npm",
      "eslint",
      "8.39.0",
    ],
    [
      "npm",
      "prettier",
      "2.8.7",
    ],
    [
      "npm",
      "typescript",
      "4.9.5",
    ],
    [
      "deno",
      "https://deno.land/x/postgres@v0.17.0/mod.ts",
      "23a1a36928f1b6d3b0170aaa67de09af12aa998525f608ff7331b9fb364cbde6",
    ],
    [
      "golang",
      "github.com/google/uuid",
      "v1.3.0",
    ],
    [
      "ruby",
      "rails",
      "6.1.7.4",
    ],
    [
      "ruby",
      "puma\\", \\"6.3.0",
      "latest",
    ],
  ],
  "edges": [],
  "id": "29",
  "inComponent": null,
  "languages": {
    "AMPL": 2,
    "Go": 3,
    "HCL": 2,
    "HTML": 3,
    "JSON": 11,
    "SCSS": 3,
    "TOML": 3,
    "YAML": 2,
  },
  "name": "flatten",
  "path": [
    "/",
    "/docker-compose.yml",
    "/deno/deno.lock",
    "/php/composer.json",
    "/golang/cmd/server/main.go",
    "/pkgs/api/package.json",
    "/pkgs/app/package.json",
    "/terraform/.terraform.lock.hcl",
    "/rust/Cargo.toml",
    "/package.json",
  ],
  "reason": [],
  "tech": null,
  "techs": [
    "algolia",
    "datadog",
    "deno",
    "docker",
    "eslint",
    "fastify",
    "gcp",
    "golang",
    "laravel",
    "nodejs",
    "php",
    "phpunit",
    "postgresql",
    "prettier",
    "prisma",
    "react",
    "redis",
    "ruby",
    "rust",
    "terraform",
    "typescript",
    "vercel",
    "vite",
  ],
}
`;
